File: src/chronoblock/syncer.py
"""
Background sync workers — one per chain, running in parallel.

Each worker loops: check the gap between the DB tip and the chain tip,
fetch the next chunk, insert, repeat. When caught up, it sleeps for
one observed block time before polling again.

Key design:
- Catch-up skips sleep and caches the chain tip to avoid one RPC call
  per chunk (~44k calls saved on a full Ethereum backfill).
- observed_block_time_ms is only recomputed after a successful sync
  that leaves us caught up, not during catch-up.
- Errors trigger a 3x backoff capped at 30s.
"""

from __future__ import annotations

import asyncio
import random
import time
from dataclasses import dataclass, field

from chronoblock import config
from chronoblock.db import (
    checkpoint_all,
    insert_blocks,
    last_block,
    observed_block_time_ms,
)
from chronoblock.errors import RpcRateLimitError, RpcResponseError
from chronoblock.log import log
from chronoblock.models import Chain
from chronoblock.rpc import fetch_block_timestamps, get_latest_block_number

__all__ = ["SyncState", "get_sync_state", "start_all", "stop_all"]

CHECKPOINT_INTERVAL = 60.0
MIN_ERROR_SLEEP = 5.0
MAX_ERROR_SLEEP = 30.0


@dataclass
class SyncState:
    started_at: float = field(default_factory=time.time)
    last_success_at: float | None = None
    last_synced_block: int | None = None
    latest_chain_block: int | None = None
    observed_block_time_ms: float = 1_000.0
    last_error: str | None = None
    last_error_at: float | None = None
    syncs_performed: int = 0
    blocks_ingested: int = 0


_sync_states: dict[int, SyncState] = {}
_tasks: list[asyncio.Task[None]] = []


def get_sync_state(chain: Chain) -> SyncState:
    state = _sync_states.get(chain.id)
    if state is not None:
        return state
    return SyncState()


# ── Lifecycle ────────────────────────────────────────────────────────


async def start_all() -> None:
    last_blocks, block_times = await asyncio.gather(
        asyncio.gather(*[asyncio.to_thread(last_block, ch) for ch in config.CHAINS]),
        asyncio.gather(*[asyncio.to_thread(observed_block_time_ms, ch) for ch in config.CHAINS]),
    )

    for chain, stored, bt in zip(config.CHAINS, last_blocks, block_times, strict=True):
        _sync_states[chain.id] = SyncState(last_synced_block=stored, observed_block_time_ms=bt)
        _tasks.append(asyncio.create_task(_sync_loop(chain), name=f"sync-{chain.name}"))

    _tasks.append(asyncio.create_task(_checkpoint_timer(), name="checkpoint-timer"))
    log("info", f"started {len(config.CHAINS)} sync workers", chain="all")


SHUTDOWN_TIMEOUT = 10.0


async def stop_all() -> None:
    for task in _tasks:
        task.cancel()
    try:
        await asyncio.wait_for(
            asyncio.gather(*_tasks, return_exceptions=True),
            timeout=SHUTDOWN_TIMEOUT,
        )
    except TimeoutError:
        log("warn", "shutdown timed out, some tasks did not exit cleanly", timeout=SHUTDOWN_TIMEOUT)
    _tasks.clear()
    await asyncio.to_thread(checkpoint_all)


# ── Sync loop ────────────────────────────────────────────────────────


async def _sync_loop(chain: Chain) -> None:
    # Stagger start so chains don't all hit their RPCs simultaneously.
    await asyncio.sleep(random.random() * 2.0)

    cached_latest: int | None = None
    consecutive_errors = 0

    while True:
        state = _sync_states[chain.id]
        try:
            catching, latest = await _sync_once(chain, cached_latest)
            state.last_error = None
            state.last_error_at = None
            state.last_success_at = time.time()
            consecutive_errors = 0

            if catching:
                cached_latest = latest
            else:
                cached_latest = None
                state.observed_block_time_ms = await asyncio.to_thread(observed_block_time_ms, chain)
                await asyncio.sleep(state.observed_block_time_ms / 1000)

        except asyncio.CancelledError:
            return
        except RpcRateLimitError as err:
            cached_latest = None
            consecutive_errors += 1
            state.last_error = str(err)
            state.last_error_at = time.time()
            escalated = min(MIN_ERROR_SLEEP * 2 ** (consecutive_errors - 1), MAX_ERROR_SLEEP)
            sleep = max(err.retry_after, escalated) if err.retry_after is not None else escalated
            log("warn", str(err), chain=chain.name, retry_in=sleep)
            await asyncio.sleep(sleep)
        except Exception as err:
            cached_latest = None
            consecutive_errors += 1
            state.last_error = str(err)
            state.last_error_at = time.time()
            sleep = min(MIN_ERROR_SLEEP * 2 ** (consecutive_errors - 1), MAX_ERROR_SLEEP)
            log("error", str(err), chain=chain.name, retry_in=sleep)
            await asyncio.sleep(sleep)


async def _sync_once(chain: Chain, cached_latest: int | None) -> tuple[bool, int]:
    latest = cached_latest if cached_latest is not None else await get_latest_block_number(chain)

    state = _sync_states[chain.id]
    state.latest_chain_block = latest

    stored = state.last_synced_block
    from_block = stored + 1 if stored is not None else 0

    if from_block > latest:
        return False, latest

    gap = latest - from_block + 1
    to_block = min(from_block + config.settings.sync_chunk_size - 1, latest)

    log("info", f"syncing {from_block}→{to_block} ({gap:,} behind)", chain=chain.name)

    blocks = await fetch_block_timestamps(chain, from_block, to_block)

    if blocks:
        await asyncio.to_thread(insert_blocks, chain, blocks)
        state.last_synced_block = blocks[-1].number
        state.syncs_performed += 1
        state.blocks_ingested += len(blocks)
    else:
        raise RpcResponseError(chain.name, f"got 0 blocks for range {from_block}→{to_block}")

    return to_block < latest, latest


async def _checkpoint_timer() -> None:
    while True:
        await asyncio.sleep(CHECKPOINT_INTERVAL)
        try:
            await asyncio.to_thread(checkpoint_all)
        except Exception as err:
            log("warn", "checkpoint failed", error=str(err))

File: typescript/syncer.ts
/**
 * @module syncer
 *
 * Background sync workers — one per chain, running in parallel.
 *
 * Each worker loops: check the gap between the DB tip and the chain
 * tip, fetch the next chunk, insert, repeat. When caught up, it
 * sleeps for one observed block time before polling again.
 *
 * The observed block time is derived from actual stored timestamps
 * (see {@link observedBlockTimeMs}), so the service self-calibrates
 * if a chain changes its block time.
 *
 * Key design decisions:
 * - Catch-up skips sleep and caches the chain tip to avoid one RPC
 *   call per chunk (~44k calls saved on a full Ethereum backfill).
 * - `observedBlockTimeMs` is only recomputed after a successful sync
 *   that leaves us caught up, never during catch-up (where it would
 *   measure historical blocks, not current ones).
 * - Errors trigger a 3× backoff capped at 30s. After a successful
 *   sync, the error backoff resets.
 */

import { CHAINS, SYNC_CHUNK_SIZE, type Chain } from "./config";
import { lastBlock, insertBlocks, checkpointAll, observedBlockTimeMs } from "./db";
import { getLatestBlockNumber, fetchBlockTimestamps } from "./rpc";
import { log } from "./log";

/** How often to truncate WAL files during sustained writes. */
const CHECKPOINT_INTERVAL_MS = 60_000;

/** Max sleep on errors, prevents runaway backoff. */
const MAX_ERROR_SLEEP_MS = 30_000;

// ── Sync state (exposed for health check + /status) ──────────────────

/** Per-chain sync state, read by the API layer for health and status. */
export interface SyncState {
  startedAt: number;
  lastSuccessAt: number | null;
  lastSyncedBlock: number | null;
  latestChainBlock: number | null;
  observedBlockTimeMs: number;
  lastError: string | null;
  lastErrorAt: number | null;
  syncsPerformed: number;
  blocksIngested: number;
}

const syncStates = new Map<number, SyncState>();

/** Returns the current sync state for a chain (never null). */
export function getSyncState(chain: Chain): SyncState {
  return syncStates.get(chain.id) ?? {
    startedAt: Date.now(),
    lastSuccessAt: null,
    lastSyncedBlock: null,
    latestChainBlock: null,
    observedBlockTimeMs: 1_000,
    lastError: null,
    lastErrorAt: null,
    syncsPerformed: 0,
    blocksIngested: 0,
  };
}

// ── Lifecycle ────────────────────────────────────────────────────────

const controllers = new Map<string, AbortController>();
const loopPromises: Promise<void>[] = [];
let checkpointTimer: ReturnType<typeof setInterval> | null = null;

/** Starts a sync worker for every configured chain. */
export function startAll() {
  for (const chain of CHAINS) {
    syncStates.set(chain.id, {
      startedAt: Date.now(),
      lastSuccessAt: null,
      lastSyncedBlock: lastBlock(chain),
      latestChainBlock: null,
      observedBlockTimeMs: observedBlockTimeMs(chain),
      lastError: null,
      lastErrorAt: null,
      syncsPerformed: 0,
      blocksIngested: 0,
    });

    const ac = new AbortController();
    controllers.set(chain.name, ac);
    loopPromises.push(syncLoop(chain, ac.signal));
  }

  checkpointTimer = setInterval(checkpointAll, CHECKPOINT_INTERVAL_MS);
  log("info", `started ${CHAINS.length} sync workers`, { chain: "all" });
}

/** Stops all sync workers, awaits loop exit, and runs a final WAL checkpoint. */
export async function stopAll() {
  if (checkpointTimer) {
    clearInterval(checkpointTimer);
    checkpointTimer = null;
  }
  for (const ac of controllers.values()) ac.abort();
  controllers.clear();
  await Promise.allSettled(loopPromises);
  loopPromises.length = 0;
  checkpointAll();
}

// ── Sync loop ────────────────────────────────────────────────────────

async function syncLoop(chain: Chain, signal: AbortSignal) {
  // Stagger start so chains don't all hit their RPCs simultaneously.
  await sleep(Math.random() * 2_000, signal);

  let cachedLatest: number | null = null;

  while (!signal.aborted) {
    const state = syncStates.get(chain.id)!;

    try {
      const { catching, latest } = await syncOnce(chain, cachedLatest, signal);
      state.lastError = null;
      state.lastErrorAt = null;
      state.lastSuccessAt = Date.now();

      if (catching) {
        // During catch-up: reuse chain tip, skip sleep and block time refresh.
        cachedLatest = latest;
      } else {
        // Caught up: refresh observed block time from fresh data, then poll.
        cachedLatest = null;
        state.observedBlockTimeMs = observedBlockTimeMs(chain);
        await sleep(state.observedBlockTimeMs, signal);
      }
    } catch (err: any) {
      if (signal.aborted) return;

      cachedLatest = null;
      state.lastError = err.message;
      state.lastErrorAt = Date.now();

      log("error", err.message, { chain: chain.name });
      await sleep(Math.min(state.observedBlockTimeMs * 3, MAX_ERROR_SLEEP_MS), signal);
    }
  }
}

/**
 * Fetches and inserts the next chunk of blocks.
 *
 * @param cachedLatest - Reuse a previously fetched chain tip to avoid
 *   an RPC round trip per chunk during catch-up.
 * @returns `catching: true` if there are more blocks to sync.
 */
async function syncOnce(
  chain: Chain,
  cachedLatest: number | null,
  signal: AbortSignal,
): Promise<{ catching: boolean; latest: number }> {
  const latest = cachedLatest ?? (await getLatestBlockNumber(chain, signal));
  const stored = lastBlock(chain);
  const from = stored !== null ? stored + 1 : 0;

  const state = syncStates.get(chain.id)!;
  state.latestChainBlock = latest;
  state.lastSyncedBlock = stored;

  if (from > latest) return { catching: false, latest };

  const gap = latest - from + 1;
  const to = Math.min(from + SYNC_CHUNK_SIZE - 1, latest);

  log("info", `syncing ${from}→${to} (${gap.toLocaleString()} behind)`, { chain: chain.name });

  const blocks = await fetchBlockTimestamps(chain, from, to, signal);

  if (blocks.length > 0) {
    insertBlocks(chain, blocks);
    state.lastSyncedBlock = blocks[blocks.length - 1].number;
    state.syncsPerformed++;
    state.blocksIngested += blocks.length;
  } else {
    throw new Error(`got 0 blocks for range ${from}→${to}`);
  }

  return { catching: to < latest, latest };
}

// ── Helpers ──────────────────────────────────────────────────────────

/** Sleep that resolves immediately when the abort signal fires. */
function sleep(ms: number, signal: AbortSignal): Promise<void> {
  if (signal.aborted) return Promise.resolve();
  return Promise.race([
    Bun.sleep(ms),
    new Promise<void>((resolve) =>
      signal.addEventListener("abort", () => resolve(), { once: true }),
    ),
  ]);
}
